Version 1.0.0

# Introduction

In this programming assignment we will illustrate a very severe data leakage, that can often be found in competitions, where the pairs of object should be scored, e.g. predict $1$ if two objects belong to the same class and $0$ otherwise. 

The data in this assignment is taken from a real competition, and the funniest thing is that *we will not use training set at all* and achieve almost 100% accuracy score! We will just exploit the leakage.

Now go through the notebook and complete the assignment.

In [1]:
import numpy as np
import pandas as pd
import scipy.sparse

# Load the data

Let's load the test data. Note, that we don't have any training data here, just test data. Moreover, *we will not even use any features* of test objects. All we need to solve this task is the file with the indices for the pairs, that we need to compare.

Let's load the data with test indices.

In [2]:
test = pd.read_csv("../readonly/data_leakages_data/test_pairs.csv")
test.head(10)

pairId  FirstId  SecondId
0       0     1427      8053
1       1    17044      7681
2       2    19237     20966
3       3     8005     20765
4       4    16837       599
5       5     3657     12504
6       6     2836      7582
7       7     6136      6111
8       8    23295      9817
9       9     6621      7672

For example, we can think that there is a test dataset of images, and each image is assigned a unique `Id` from $0$ to $N-1$ (N -- is the number of images). In the dataframe from above `FirstId` and `SecondId` point to these `Id`'s and define pairs, that we should compare: e.g. do both images in the pair belong to the same class or not. So, for example for the first row: if images with `Id=1427` and `Id=8053` belong to the same class, we should predict $1$, and $0$ otherwise. 

But in our case we don't really care about the images, and how exactly we compare the images (as long as comparator is binary).  

**We suggest you to try to solve the puzzle yourself first.** You need to submit a `.csv` file with columns `pairId` and `Prediction` to the grader. The number of submissions allowed is made pretty huge to let you explore the data without worries. The returned score should be very close to $1$.

**If you do not want to think much** -- scroll down and follow the instructions below.

In [3]:
test.describe()

pairId        FirstId       SecondId
count  368550.000000  368550.000000  368550.000000
mean   184274.500000   10863.601118   11950.398882
std    106391.365192    7280.190939    7602.814820
min         0.000000       0.000000       0.000000
25%     92137.250000    4574.000000    5698.000000
50%    184274.500000    9886.000000   10512.000000
75%    276411.750000   16512.000000   18782.000000
max    368549.000000   26324.000000   26324.000000

In [4]:
len(test.pairId.unique())

368550

In [5]:
len(test.FirstId.unique())

26325

In [6]:
len(test.SecondId.unique())

26310

# EDA and leakage intuition

As we already know, the key to discover data leakages is careful EDA. So let's start our work with some basic data exploration and build an intuition about the leakage.

First, check, how many different `id`s are there: concatenate `FirstId` and `SecondId` and print the number of unique elements. Also print minimum and maximum value for that vector.

In [18]:
# YOUR CODE GOES HERE

total_item = pd.concat([test.FirstId, test.SecondId], ignore_index=True).unique()
print("total ids:", len(total_item))
print("max id:", max(total_item))

total ids: 26325
max id: 26324


and then print how many pairs we need to classify (it is basically the number of rows in the test set)

In [19]:
# YOUR CODE GOES HERE
print("len of test:", len(test))

len of test: 368550


Now print, how many distinct pairs it would be possible to create out of all "images" in the dataset?   

In [20]:
# YOUR CODE GOES HERE
print("Possible pairs:", len(total_item) * (len(total_item) - 1))

Possible pairs: 692979300


So the number of pairs we are given to classify is very very small compared to the total number of pairs. 

To exploit the leak we need to **assume (or prove)**, that the total number of positive pairs is small, compared to the total number of pairs. For example: think about an image dataset with $1000$ classes, $N$ images per class. Then if the task was to tell whether a pair of images belongs to the same class or not, we would have $1000\frac{N(N-1)}{2}$ positive pairs, while total number of pairs was $\frac{1000N(1000N - 1)}{2}$.

Another example: in [Quora competitition](https://www.kaggle.com/c/quora-question-pairs) the task was to classify whether a pair of qustions are duplicates of each other or not. Of course, total number of question pairs is very huge, while number of duplicates (positive pairs) is much much smaller.

Finally, let's get a fraction of pairs of class `1`. We just need to submit a constant prediction "all ones" and check the returned accuracy. Create a dataframe with columns `pairId` and `Prediction`, fill it and export it to `.csv` file. Then submit to grader and examine grader's output. 

In [10]:
# YOUR CODE GOES HERE

submission = test.copy()
submission["Prediction"] = 1
submission.drop(columns=["FirstId", "SecondId"])
submission.to_csv("./all_one.csv", index=False, encoding="utf-8")

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

1. 正負資料是 1:1
2. FistId 有所有的 image

In [30]:
int(len(test) * 0.5)

184275

In [48]:
test.head(20)

pairId  FirstId  SecondId
0        0     1427      8053
1        1    17044      7681
2        2    19237     20966
3        3     8005     20765
4        4    16837       599
5        5     3657     12504
6        6     2836      7582
7        7     6136      6111
8        8    23295      9817
9        9     6621      7672
10      10    14913     25036
11      11    10182     21628
12      12     7891     15658
13      13    20350      3568
14      14     3101      9799
15      15    16381      1565
16      16    16564      7991
17      17     5279      8834
18      18    13301     10079
19      19     3376     24330

In [40]:
pred1 = [0 for x in range(int(len(test) * 0.5))]
pred2 = [1 for x in range(int(len(test) * 0.5))]
pred = pred1 + pred2

submission = test.copy()
submission["Prediction"] = pred
submission = submission.drop(columns=["FirstId", "SecondId"])
submission.to_csv("./zero_one.csv", index=False, encoding="utf-8")
submission

pairId  Prediction
0            0           0
1            1           0
2            2           0
3            3           0
4            4           0
...        ...         ...
368545  368545           1
368546  368546           1
368547  368547           1
368548  368548           1
368549  368549           1

[368550 rows x 2 columns]

In [85]:
duplicate = test[test.duplicated(subset=["FirstId", "SecondId"])]
duplicate

pairId  FirstId  SecondId
107647  107647    10008     10404
123152  123152     3505      8046
124261  124261     5641      8046
139487  139487      504      2799
226778  226778     1980      8046
238033  238033     1420      8046
239980  239980     2627     10404
256914  256914      332     10404
280690  280690     6957     10404
310044  310044     6291     10404
330339  330339     4155     10404
353199  353199      256      2551

In [89]:
test[(test["FirstId"] == 10008) & (test["SecondId"] == 10404)]

pairId  FirstId  SecondId
107364  107364    10008     10404
107647  107647    10008     10404

In [21]:
# 保證是正資料的都在下半部 (巧合)
test[test["FirstId"] == test["SecondId"]]

pairId  FirstId  SecondId
186443  186443      300       300
218927  218927     2551      2551
221189  221189     8046      8046
228783  228783     2799      2799
293834  293834      292       292
327330  327330    10404     10404

# Building a magic feature

In this section we will build a magic feature, that will solve the problem almost perfectly. The instructions will lead you to the correct solution, but please, try to explain the purpose of the steps we do to yourself -- it is very important.

## Incidence matrix

First, we need to build an [incidence matrix](https://en.wikipedia.org/wiki/Incidence_matrix). You can think of pairs `(FirstId, SecondId)` as of edges in an undirected graph. 

The incidence matrix is a matrix of size `(maxId + 1, maxId + 1)`, where each row (column) `i` corresponds `i-th` `Id`. In this matrix we put the value `1` to the position `[i, j]`, if and only if a pair `(i, j)` or `(j, i)` is present in  a given set of pais `(FirstId, SecondId)`. All the other elements in the incidence matrix are zeros.   

**Important!** The incidence matrices are typically very very sparse (small number of non-zero values). At the same time incidence matrices are usually huge in terms of total number of elements, and it is **impossible to store them in memory in dense format**. But due to their sparsity incidence matrices **can be easily represented as sparse matrices**. If you are not familiar with sparse matrices, please see [wiki](https://en.wikipedia.org/wiki/Sparse_matrix) and [scipy.sparse reference](https://docs.scipy.org/doc/scipy/reference/sparse.html). Please, use any of `scipy.sparse` constructors to build incidence matrix. 

For example, you can use this constructor: `scipy.sparse.coo_matrix((data, (i, j)))`. We highly recommend to learn to use different `scipy.sparse` constuctors, and matrices types, but if you feel you don't want to use them, you can always build this matrix with a simple `for` loop. You will need first to create a matrix using `scipy.sparse.coo_matrix((M, N), [dtype])` with an appropriate shape `(M, N)` and then iterate through `(FirstId, SecondId)` pairs and fill corresponding elements in matrix with ones. 

**Note**, that the matrix should be symmetric and consist only of zeros and ones. It is a way to check yourself.

In [63]:
test

pairId  FirstId  SecondId
0            0     1427      8053
1            1    17044      7681
2            2    19237     20966
3            3     8005     20765
4            4    16837       599
...        ...      ...       ...
368545  368545     2996      6659
368546  368546    13626      3390
368547  368547    25407      2932
368548  368548    25305      5588
368549  368549     1767     11874

[368550 rows x 3 columns]

In [118]:
duplicate = test[test.duplicated(subset=["FirstId", "SecondId"])]
duplicate.index

Int64Index([107647, 123152, 124261, 139487, 226778, 238033, 239980, 256914,
            280690, 310044, 330339, 353199],
           dtype='int64')

In [119]:
test2 = test.copy()
test2 = test2.drop(duplicate.index)
test2

pairId  FirstId  SecondId
0            0     1427      8053
1            1    17044      7681
2            2    19237     20966
3            3     8005     20765
4            4    16837       599
...        ...      ...       ...
368545  368545     2996      6659
368546  368546    13626      3390
368547  368547    25407      2932
368548  368548    25305      5588
368549  368549     1767     11874

[368538 rows x 3 columns]

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

In [124]:
i = np.concatenate((test.FirstId.values, test.SecondId.values), axis=0)
j = np.concatenate((test.SecondId.values, test.FirstId.values), axis=0)

aa = pd.DataFrame({"FirstId": i, "SecondId": j})
aa

FirstId  SecondId
0          1427      8053
1         17044      7681
2         19237     20966
3          8005     20765
4         16837       599
...         ...       ...
737095     6659      2996
737096     3390     13626
737097     2932     25407
737098     5588     25305
737099    11874      1767

[737100 rows x 2 columns]

In [128]:
duplicate = aa[aa.duplicated(subset=["FirstId", "SecondId"])]
duplicate

FirstId  SecondId
107647    10008     10404
123152     3505      8046
124261     5641      8046
139487      504      2799
226778     1980      8046
...         ...       ...
734226     6463      2799
734412      292      4115
735255     8046     20806
735456     8736      9417
735801    11083      2487

[228 rows x 2 columns]

In [130]:
aa = aa.drop(duplicate.index)
aa

FirstId  SecondId
0          1427      8053
1         17044      7681
2         19237     20966
3          8005     20765
4         16837       599
...         ...       ...
737095     6659      2996
737096     3390     13626
737097     2932     25407
737098     5588     25305
737099    11874      1767

[736872 rows x 2 columns]

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

In [131]:
i = aa.FirstId.values
j = aa.SecondId.values
data = np.ones(len(i))
inc_mat = scipy.sparse.coo_matrix(
    (data, (i, j)), shape=(26325, 26325)
)  # YOUR CODE GOES HERE (but probably you will need to write few more lines before)

# Sanity checks
assert inc_mat.max() == 1
assert inc_mat.sum() == 736872

It is convenient to have matrix in `csr` format eventually.

In [132]:
inc_mat = inc_mat.tocsr()

## Now build the magic feature

Why did we build the incidence matrix? We can think of the rows in this matix as of representations for the objects. `i-th` row is a representation for an object with `Id = i`. Then, to measure similarity between two objects we can measure similarity between their representations. And we will see, that such representations are very good.

Now select the rows from the incidence matrix, that correspond to `test.FirstId`'s, and `test.SecondId`'s.

In [144]:
# Note, scipy goes crazy if a matrix is indexed with pandas' series.
# So do not forget to convert `pd.series` to `np.array`
# These lines should normally run very quickly

rows_FirstId = inc_mat[test.FirstId.values]  # YOUR CODE GOES HERE
rows_SecondId = inc_mat[test.SecondId.values]  # YOUR CODE GOES HERE

Our magic feature will be the *dot product* between representations of a pair of objects. Dot product can be regarded as similarity measure -- for our non-negative representations the dot product is close to 0 when the representations are different, and is huge, when representations are similar. 

Now compute dot product between corresponding rows in `rows_FirstId` and `rows_SecondId` matrices.

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

In [236]:
# Note, that in order to do pointwise multiplication in scipy.sparse you need to use function `multiply`
# regular `*` corresponds to matrix-matrix multiplication

f = rows_FirstId.multiply(rows_SecondId).sum(axis=1)  # YOUR CODE GOES HERE
f = np.asarray(f).reshape(-1)

# Sanity check
assert f.shape == (368550,)

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

That is it! **We've built our magic feature.** 

# From magic feature to binary predictions

But how do we convert this feature into binary predictions? We do not have a train set to learn a model, but we have a piece of information about test set: the baseline accuracy score that you got, when submitting constant. And we also have a very strong considerations about the data generative process, so probably we will be fine even without a training set. 

We may try to choose a thresold, and set the predictions to 1, if the feature value `f` is higer than the threshold, and 0 otherwise. What threshold would you choose? 

How do we find a right threshold? Let's first examine this feature: print frequencies (or counts) of each value in the feature `f`.

In [250]:
# For example use `np.unique` function, check for flags

# YOUR CODE GOES HERE
unique, counts = np.unique(f, return_counts=True)
dict(zip(unique, counts))

{14.0: 183279, 15.0: 852, 19.0: 546, 20.0: 183799, 21.0: 6, 28.0: 54, 35.0: 14}

Do you see how this feature clusters the pairs? Maybe you can guess a good threshold by looking at the values? 

In fact, in other situations it can be not that obvious, but in general to pick a threshold you only need to remember the score of your baseline submission and use this information. Do you understand why and how?  

Choose a threshold below: 

In [251]:
pred = f > 15  # SET THRESHOLD HERE

In [252]:
pred

array([ True, False,  True, ..., False, False, False])

In [255]:
from collections import Counter

Counter(pred).keys()  # equals to list(set(words))

dict_keys([True, False])

In [256]:
Counter(pred).values()  # counts the elements' frequency

dict_values([184419, 184131])

# Finally, let's create a submission

In [257]:
submission = test.loc[:, ["pairId"]]
submission["Prediction"] = pred.astype(int)

submission.to_csv("submission.csv", index=False)

Now submit it to the grader! It is not possible to submit directly from this notebook, as we need to submit a `csv` file, not a single number (limitation of Coursera platform). 

To download `submission.csv` file that you've just produced <a href='./submission.csv'>click here</a> (if the link opens in browser, right-click on it and shoose "Save link as"). Then go to [assignment page](https://www.coursera.org/learn/competitive-data-science/programming/KsASv/data-leakages/submission) and submit your `.csv` file in 'My submission' tab.


If you did everything right, the score should be very high.

**Finally:** try to explain to yourself, why the whole thing worked out. In fact, there is no magic in this feature, and the idea to use rows in the incidence matrix can be intuitively justified.

# Bonus

Interestingly, it is not the only leak in this dataset. There is another totally different way to get almost 100% accuracy. Try to find it!